### PR TITLE
MAILET-146 Add a mailet to attach a parsed Calendar to mail HEADERs

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+import net.fortuna.ical4j.model.Calendar;
+import net.fortuna.ical4j.model.Property;
+import net.fortuna.ical4j.model.component.VEvent;
+
+public class ICALToHeader extends GenericMailet {
+    public static final String ATTRIBUTE_PROPERTY = "attribute";
+    public static final String ATTRIBUTE_DEFAULT_NAME = "icalendar";
+
+    public static final String X_MEETING_UID = "X-MEETING-UID";
```
Comment (rouazana) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: _HEADER for all of these?

File: mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        try {
+            Map<String, Calendar> calendars = (Map<String, Calendar>) mail.getAttribute(attribute);
+            Optional<Calendar> calendar = calendars.values()
+                .stream()
+                .findAny();
+            if (calendar.isPresent()) {
+                writeToHeaders(calendar.get(), mail.getMessage());
+            }
+        } catch (ClassCastException e) {
+            LOGGER.error("Received a mail with " + attribute + " not being and ICAL object", e);
```
Comment (rouazana) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: s/and/an/ + add an identifier for the mail?

File: mailet/icalendar/src/test/java/org/apache/james/transport/mailets/ICALToHeadersTest.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            .containsOnly("f1514f44bf39311568d640727cff54e819573448d09d2e5677987ff29caa01a9e047feb2aab16e43439a608f28671ab7c10e754ce92be513f8e04ae9ff15e65a9819cf285a6962bc");
+        assertThat(mail.getMessage().getHeader(ICALToHeader.X_MEETING_DTSTAMP)).containsOnly("20170106T115036Z");
+        assertThat(mail.getMessage().getHeader(ICALToHeader.X_MEETING_RECURRENCE_ID)).isNull();
+        assertThat(mail.getMessage().getHeader(ICALToHeader.X_MEETING_SEQUENCE)).containsOnly("0");
+    }
+
+    @Test
+    public void serviceShouldWriteOnlyOneICalendarToHeaders() throws Exception {
+        Calendar calendar = new CalendarBuilder().build(ClassLoader.getSystemResourceAsStream("ics/meeting.ics"));
+        Calendar calendar_2 = new CalendarBuilder().build(ClassLoader.getSystemResourceAsStream("ics/meeting.ics"));
```
Comment (rouazana) [inline mailet/icalendar/src/test/java/org/apache/james/transport/mailets/ICALToHeadersTest.java]: _ in variable?

Comment (rouazana) [inline mailet/icalendar/src/test/java/org/apache/james/transport/mailets/ICALToHeadersTest.java]: you don't use meeting_2

File: mailet/icalendar/src/test/java/org/apache/james/transport/mailets/ICALToHeadersTest.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        Mail mail = FakeMail.builder()
+            .mimeMessage(MimeMessageBuilder.defaultMimeMessage())
+            .attribute(ICALToHeader.ATTRIBUTE_DEFAULT_NAME, icals)
+            .build();
+
+        testee.service(mail);
+
+        assertThat(mail.getMessage().getHeader(ICALToHeader.X_MEETING_UID)).isNull();
+    }
+}
```
Comment (rouazana) [inline mailet/icalendar/src/test/java/org/apache/james/transport/mailets/ICALToHeadersTest.java]: can you add a test with some empty headers?

Comment (chibenwa) [inline mailet/icalendar/src/test/java/org/apache/james/transport/mailets/ICALToHeadersTest.java]: I don't understand this one...

Comment (rouazana) [inline mailet/icalendar/src/test/java/org/apache/james/transport/mailets/ICALToHeadersTest.java]: some ics doesn't contain all the data you expect, and then you'll not add these headers. Please add a test showing this

File: mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    @Override
+    public String getMailetInfo() {
+        return "ICALToHeader Mailet";
+    }
+
+    @Override
+    public void init() throws MessagingException {
+        attribute = getInitParameter(ATTRIBUTE_PROPERTY, ATTRIBUTE_DEFAULT_NAME);
+        if (Strings.isNullOrEmpty(attribute)) {
+            throw new MessagingException("attribute can not be empty or null");
```
Comment (mbaechler) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: you should put the attribute name here

File: mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        }
+    }
+
+    @SuppressWarnings("unchecked")
```
Comment (mbaechler) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: could you extract the unchecked part in a smaller method ?

File: mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    public void service(Mail mail) throws MessagingException {
+        if (mail.getAttribute(attribute) == null) {
+            return;
+        }
+        try {
+            Map<String, Calendar> calendars = (Map<String, Calendar>) mail.getAttribute(attribute);
+            Optional<Calendar> calendar = calendars.values()
+                .stream()
+                .findAny();
+            if (calendar.isPresent()) {
```
Comment (mbaechler) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: useless isPresent IMO, you should call the method in the stream

File: mailet/icalendar/src/test/resources/ics/meeting_2.ics [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+ com
+X-OBM-DOMAIN:linagora.com
+X-OBM-DOMAIN-UUID:02874f7c-d10e-102f-acda-0015176f7922
+LOCATION:hangout
+CATEGORIES:
+X-OBM-COLOR:
+UID:f1514f44bf39311568d64072ac247c17656ceafde3b4b3eba961c8c5184cdc6ee047fe
+ b2aab16e43439a608f28671ab7c10e754c301b1e32001ad51dd20eac2fc7af20abf4093bbe
+ATTENDEE;CUTYPE=INDIVIDUAL;RSVP=TRUE;CN=Michael BAILLY;PARTSTAT=NEEDS-ACTI
+ ON;X-OBM-ID=348:MAILTO:mbailly@linagora.com
```
Comment (mbaechler) [inline mailet/icalendar/src/test/resources/ics/meeting_2.ics]: please anonymize

File: mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        addIfPresent(mimeMessage, X_MEETING_DTSTAMP_HEADER, vevent.getDateStamp());
+        mimeMessage.saveChanges();
+    }
+
+    private void addIfPresent(MimeMessage mimeMessage, String headerName, Property property) {
+        if (property != null) {
+            try {
+                mimeMessage.addHeader(headerName, property.getValue());
+            } catch (MessagingException e) {
+                LOGGER.warn("Could not add header {} with value {}", headerName, property.getValue());
```
Comment (aduprat) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: Log with the GenericMailet.log method.

Comment (chibenwa) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: I really don't like it. I can log with getLogger() if you prefer.

Comment (aduprat) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: Me too, but we have to use the same mechanism in each mailets; then it will be easier to refactor, IMHO.

Comment (chibenwa) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: To explain this one a bit better, I really think centralizing mailet logging thew the MailetContainer (Akka log() calls of Generic mailets) will make fined grain configuration of logging (using class names for instance) will be wayharder as an admin.

Concerning the log method of the generic mailet, I don't see why I would have to use a poor Logging API, without placehorlder, and were you will have to guess the log level of what you are doing.

For older mailets, maybe we can not change it for now but for new mailets, maybe we should go away from this API

Comment (chibenwa) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: I used log from GenericMailet instead

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

File: mailet/icalendar/src/test/java/org/apache/james/transport/mailets/ICALToHeadersTest.java [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    @Test
+    public void serviceShouldNotModifyMailsWithoutIcalAttribute() throws Exception {
+        testee.init(FakeMailetConfig.builder().build());
+        Mail mail = FakeMail.builder()
+            .mimeMessage(MimeMessageBuilder.defaultMimeMessage())
+            .build();
+
+        testee.service(mail);
+
+        assertThat(mail.getAttributeNames()).isEmpty();
```
Comment (rouazana) [inline mailet/icalendar/src/test/java/org/apache/james/transport/mailets/ICALToHeadersTest.java]: It seems you are not checking that you are not modifying the mail content?

File: mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        addIfPresent(mimeMessage, X_MEETING_DTSTAMP_HEADER, vevent.getDateStamp());
+        mimeMessage.saveChanges();
+    }
+
+    private void addIfPresent(MimeMessage mimeMessage, String headerName, Property property) {
+        if (property != null) {
+            try {
+                mimeMessage.addHeader(headerName, property.getValue());
+            } catch (MessagingException e) {
+                log("Could not add header " + headerName + " with value " +property.getValue(), e);
```
Comment (rouazana) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: one space missing

File: mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    private void writeToHeaders(Calendar calendar, MimeMessage mimeMessage) throws MessagingException {
+        VEvent vevent = (VEvent) calendar.getComponent("VEVENT");
+        addIfPresent(mimeMessage, X_MEETING_METHOD, calendar.getMethod());
+        addIfPresent(mimeMessage, X_MEETING_UID, vevent.getUid());
+        addIfPresent(mimeMessage, X_MEETING_RECURRENCE_ID, vevent.getRecurrenceId());
+        addIfPresent(mimeMessage, X_MEETING_SEQUENCE, vevent.getSequence());
+        addIfPresent(mimeMessage, X_MEETING_DTSTAMP, vevent.getDateStamp());
+        mimeMessage.saveChanges();
```
Comment (aduprat) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: Please remove this call, has it's already handled by lower level method calls (MimeMessage.writeTo).
NB: i think we can definitively remove all calls of this method, has it make sense only for folder backend.

**Comment by chibenwa** [comment]
First one, last one, whatever. How do you define the ordering?

**Comment by chibenwa** [comment]
What's important is having a single attachement into the headers IMO

**Comment by Laura-Royet** [comment]
First ICS attachment is the first ICS attachment you join to your email etc

**Comment by chibenwa** [comment]
The notion of order in a tree like structure, after a conversion in a map is kind of tricky and IMO do not bring any business value to your test case. Just assert that you get one.

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

File: mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java (hunk 12 above)
Comment (chibenwa) [inline mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICALToHeader.java]: Knew it ;-)

Agree, I remove it here. Let's see to remove it in other places later

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Concider it green : CI failed to notify success...

```
Caused by: java.net.ConnectException: Connection timed out
	at sun.reflect.GeneratedConstructorAccessor332.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1889)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1884)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.net.www.protocol.http.HttpURLConnection.getChainedException(HttpURLConnection.java:1883)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1456)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1440)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:338)
	at org.kohsuke.github.Requester.handleApiError(Requester.java:489)
	at org.kohsuke.github.Requester._to(Requester.java:248)
	at org.kohsuke.github.Requester.to(Requester.java:194)
```

**Comment by aduprat** [comment]
Build finished. No test results found.

**Comment by chibenwa** [comment]
test this please
